sherpa/cpp_api/websocket/client.cc

```diff
@@ -0,0 +1,137 @@
+/**
+ * Copyright      2022  (authors: Pingfeng Luo)
+ *
+ * See LICENSE for clarification regarding multiple authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include <string>
+#include <thread> // NOLINT
+
+#include "boost/asio/connect.hpp"
+#include "boost/asio/ip/tcp.hpp"
+#include "boost/beast/core.hpp"
+#include "boost/beast/websocket.hpp"
+#include "boost/json/src.hpp"
+
+#include "sherpa/csrc/fbank_features.h"
+#include "sherpa/csrc/log.h"
+#include "sherpa/csrc/parse_options.h"
+
+
+namespace asio = boost::asio;
+using tcp = boost::asio::ip::tcp;
+namespace beast = boost::beast;
+namespace json = boost::json;
+namespace http = beast::http;
+namespace websocket = beast::websocket;
+
+class WebSocketClient {
+ public:
+  WebSocketClient(const std::string& hostname, int port)
+    : hostname_(hostname), port_(port) {
+      Connect();
+      t_.reset(new std::thread(&WebSocketClient::ReadLoopFunc, this));
+    }
+
+  void SendBinaryData(const void* data, size_t size) {
+    ws_.binary(true);
+    ws_.write(asio::buffer(data, size));
+  }
+
+  void ReadLoopFunc() {
+    try {
+      while (true) {
+        beast::flat_buffer buffer;
+        ws_.read(buffer);
+        std::string message = beast::buffers_to_string(buffer.data());
+        SHERPA_LOG(INFO) << message;
+        json::object obj = json::parse(message).as_object();
+        if (obj["status"] != "ok") {
+          break;
+        }
+      }
+    } catch (beast::system_error const& se) {
```

Comment:
```suggestion
    } catch (const beast::system_error& se) {
```

To be consistent with other code in sherpa, we use
```
const some_type& var;
```